### PR TITLE
Use checksums from traefik releases

### LINKED ISSUE #146: automate updating hashes of binaries in install.py
### Proposed change

It would be nice if updating the hashes of the installable binaries could be automated. This will make it easier to support more platforms (e.g. arm) and update versions, which will get tedious fast if we have to do them all by hand.

### Alternative options

keep doing it by hand

### Who would use this feature?

maintainers who are updating the supported version(s) of traefik, etc. which can be fetched.

### (Optional): Suggest a solution

store hashes in a json or toml file instead of Python, and then have a script that

1. scrapes releases for versions,
2. retrieves (or calculates) hashes, and
3. writes an updated file

Install.py would only change in that the currently hardcoded dicts become a read from a file.

Bonus points if it's run in a github workflow that issues PRs itself, but running manually and opening a PR works just fine.

### LINKED ISSUE #147: update bundled traefik
#145 gets our traefik v2 support, but the v2 versions are quire out of date. We should update to latest versions of the bundled dependencies once it's working

### PR DESCRIPTION
**Pull request by minrk**
install.py now uses the official checksums from traefik releases to verify downloads.

This means checksums work for any published traefik release, not just whatever we have downloaded.

Technically, this allows someone to publish a malicious release to traefik itself and we wouldn't catch it (assuming they checksum their malicious files). But our own checksum-updating wouldn't catch the same thing anyway, as it requires compromising the traefik release-publishing pipeline and/or GitHub (or HTTPS) itself.

I think for the scale of robustness vs convenience, this is the right level for us. Folks are always free to take a more rigorous path for their own deployments.

closes #146

Default Traefik is updated to 2.9.8 (closes #147)

Original PR:

<del>
`./ci/update_hashes.py add` will always add the hashes for the latest traefik release, so we shouldn't need to compute any hashes or URLs by hand.

I've updated it to include the latest patch version of the latest two minor releases of traefik.
</del>

### CODE REVIEW

File: jupyterhub_traefik_proxy/install.py [hunk 1]
```diff
@@ -2,26 +2,59 @@
...
 import sys
 import tarfile
 import textwrap
 import warnings
 import zipfile
 from urllib.request import urlretrieve
 
+from packaging.version import parse as parse_version
+
+# BEGIN CHECKSUMS
```
Comment (GeorgianaElena) [inline jupyterhub_traefik_proxy/install.py:15]: Courious about the choice of keeping this in `install.py` and not into their own.

It makes me a little nervous knowing the `update_hashes.py` script modifies this entire file and that an accidental delition of  the `# BEGIN CHECKSUMS` comments will break things.

**Review by GeorgianaElena — APPROVED** [review]
> , so we shouldn't need to compute any hashes or URLs by hand.

This is super nice! Thanks @minrk

I left a tiny comment about choosing to keep the hashes in the installer, but I don't think that should block merging this PR

File: jupyterhub_traefik_proxy/install.py (hunk 1 above)
Comment (minrk) [reply jupyterhub_traefik_proxy/install.py:15]: I thought about that! We could definitely put it in its own file, that would make some things easier.

**Comment by minrk** [comment]
@GeorgianaElena It's updated to now populate a standard `checksums.txt` that is essentially a concatenation of traefik's own published checksum files.

That being the case, I'd also be okay with removing the script and going back to manual, since it can now be done with a single copy/paste from the checksums.txt at https://github.com/traefik/traefik/releases. Ultimately, it's:

    V=2.9.8
    curl -v -L https://github.com/traefik/traefik/releases/download/v${V}/traefik_v${V}_checksums.txt >> checksums.txt; cat checksums.txt | uniq

**Comment by GeorgianaElena** [comment]
Hmm, now that traefik publishes the checksums itself, I'm now wondering why keep a `checksums.txt` ourselves at all, and not check the checksums directly from the installer given a provided version.

I'm assuming it's to ensure and assure a certain compatibility, but couldn't just be checked in our tests on the CI?

Maybe `traefik.installer.py` doesn't make so much sense  and instead we should just have this be a helper script to use in the CI and not have it be part of the package at all?

**Comment by minrk** [comment]
So, I think it's a bit silly. But the one thing our checksum copy protects against is the traefik GitHub releases being compromised (or https to GitHub itself!). And the same thing is technically true at the time when we collect the checksums. Essentially, it verifies that these things don't change _over time_ without us knowing about it.

That's _extremely_ unlikely, so I'd be happy to fetch the checksum as part of the download process itself, and get rid of our own tracking.

**Comment by minrk** [comment]
I do think being able to fetch and install the traefik binary is sufficiently useful that it's worth keeping in the package.

**Comment by minrk** [comment]
@GeorgianaElena I've updated this to now just use the official traefik checksums from releases, and removed any need to track our own checksum copies

**Comment by GeorgianaElena** [comment]
Thank you, @minrk! I'm +1 with the resolution here so let's go for a merge.

**Pull request merged**